src/storage/v2/indices/text_index.cpp

```diff
@@ -0,0 +1,470 @@
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  auto applicable_text_indices = GetApplicableTextIndices(vertex_after_update);
+  if (applicable_text_indices.empty()) return;
+  RemoveNode(vertex_after_update, applicable_text_indices);
+  AddNode(vertex_after_update, storage, transaction_start_timestamp, applicable_text_indices);
+}
+
+void TextIndex::UpdateNode(Vertex *vertex_after_update, Storage *storage,
+                           const std::uint64_t transaction_start_timestamp,
+                           const std::vector<LabelId> &removed_labels) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  auto indexes_to_remove_node_from = GetApplicableTextIndices(removed_labels);
+  RemoveNode(vertex_after_update, indexes_to_remove_node_from);
+
+  auto indexes_to_update_node = GetApplicableTextIndices(vertex_after_update);
+  if (indexes_to_update_node.empty()) return;
+  RemoveNode(vertex_after_update, indexes_to_update_node);
+  AddNode(vertex_after_update, storage, transaction_start_timestamp, indexes_to_update_node);
+}
+
+void TextIndex::RemoveNode(Vertex *vertex_after_update,
+                           const std::vector<mgcxx::text_search::Context *> &applicable_text_indices) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  auto search_node_to_be_deleted =
+      mgcxx::text_search::SearchInput{.search_query = fmt::format("metadata.gid:{}", vertex_after_update->gid.AsInt())};
+
+  for (auto *index_context : applicable_text_indices) {
+    try {
+      mgcxx::text_search::delete_document(*index_context, search_node_to_be_deleted, kDoSkipCommit);
+    } catch (const std::exception &e) {
+      throw query::QueryException(fmt::format("Tantivy error: {}", e.what()));
+    }
+  }
+}
+
+void TextIndex::RemoveNode(Vertex *vertex_after_update) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  auto applicable_text_indices = GetApplicableTextIndices(vertex_after_update);
+  if (applicable_text_indices.empty()) return;
+  RemoveNode(vertex_after_update, applicable_text_indices);
+}
+
+void TextIndex::UpdateOnAddLabel(LabelId added_label, Vertex *vertex_after_update, Storage *storage,
+                                 const std::uint64_t transaction_start_timestamp) {
+  if (!flags::run_time::GetExperimentalTextSearchEnabled()) {
+    throw query::QueryException("To use text indices, enable the text search feature.");
+  }
+
+  if (!label_to_index_.contains(added_label)) {
+    return;
+  }
+  AddNode(vertex_after_update, storage, transaction_start_timestamp,
```

Comment:
Can you maybe explain on AddLabel we call AddNode?

Reply:
Text indices are defined per-label. Say there’s a text index on `:Document`. When we add that label to a node, that node has to be added to the text index.